src/lib/homestore_backend/hs_homeobject.cpp

```diff
@@ -87,16 +86,40 @@ HSHomeObject::~HSHomeObject() {
     iomanager.stop();
 }
 
-void HSHomeObject::on_shard_meta_blk_found(homestore::meta_blk* mblk, sisl::byte_view buf, size_t size) {
-    std::string shard_info_str;
-    shard_info_str.append(r_cast< const char* >(buf.bytes()), size);
+void HSHomeObject::on_shard_meta_blk_found(homestore::meta_blk* mblk, sisl::byte_view buf) {
+    homestore::superblk< shard_info_superblk > sb;
+    sb.load(buf, mblk);
 
-    auto shard = deserialize_shard(shard_info_str);
-    shard.metablk_cookie = mblk;
+    bool pg_is_recovered = false;
+    {
+        std::scoped_lock lock_guard(_pg_lock);
+        pg_is_recovered = _pg_map.find(sb->placement_group) != _pg_map.end();
+    }
 
-    // As shard info in the homestore metablk is always the latest state(OPEN or SEALED),
-    // we can always create a shard from this shard info and once shard is deleted, the associated metablk will be
-    // deleted too.
-    do_commit_new_shard(shard);
+    if (pg_is_recovered) {
+        add_new_shard_to_map(std::make_unique< HS_Shard >(sb));
+        return;
+    }
+
+    // There is no guarantee that pg info will be recovered before shard recovery
+    std::scoped_lock lock_guard(recovery_mutex_);
```

Comment:
IIRC, the callbacks of metaservice will not be called concurrently, but sequentially in the function(https://github.com/eBay/HomeStore/blob/bf8e6fdf9a1a417dd5aaf7be7328b4d015dd1caa/src/lib/meta/meta_blk_service.cpp#L1074). so , do we need this lock here